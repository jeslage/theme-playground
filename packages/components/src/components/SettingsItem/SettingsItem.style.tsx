import styled from 'styled-components';

export const StyledSettingsItem = styled.div`
  padding: 0.75em 1rem;
  min-height: 50px;
  display: flex;
  align-items: center;
  justify-content: space-between;

  &:nth-of-type(odd) {
    background: ${props => props.theme.backgroundLight};
  }
`;
