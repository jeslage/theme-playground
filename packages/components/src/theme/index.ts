export const darkTheme = {
  base: 'dark',
  backgroundDark: '#2f2f2f',
  backgroundLight: '#333',
  border: '#999999',
  text: '#FFFFFF'
};

export const lightTheme = {
  base: 'light',
  backgroundDark: '#f6f9fc',
  backgroundLight: '#fff',
  border: '#666666',
  text: '#333333'
};

export const getUiTheme = theme => {
  if (theme === undefined) return darkTheme;

  if (typeof theme === 'string') {
    if (theme === 'dark') {
      return darkTheme;
    } else if (theme === 'light') {
      return lightTheme;
    }

    throw Error(
      `Theme "${theme}" is not available. Please use "light" or "dark", or provide a custom theme object.`
    );
  }

  if (theme.base && theme.base === 'light') {
    return {
      ...lightTheme,
      ...theme
    };
  }

  return {
    ...darkTheme,
    ...theme
  };
};
