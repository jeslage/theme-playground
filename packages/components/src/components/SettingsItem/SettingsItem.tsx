import * as React from 'react';

import { ThemePlaygroundContext } from '../../contexts/ThemePlaygroundProvider';

import { getLabel } from '../../helper';

import Colorpicker from '../ColorPicker/ColorPicker';
import Counter from '../Counter/Counter';
import Input from '../Input/Input';
import RadioGroup from '../RadioOption/RadioOption';
import Range from '../Range/Range';
import Select from '../Select/Select';
import Shorthand from '../Shorthand/Shorthand';
import Switch from '../Switch/Switch';
import Textarea from '../Textarea/Textarea';

import { StyledSettingsItem } from './SettingsItem.style';
import { stripUnit } from '../../helper/stripUnit';

interface ComponentProps {
  type: string;
  path: string;
  props: { label: string; value: any };
  overrideProps: any;
  update: (path: string, value: any) => void;
}

export const SettingsComponent: React.FC<ComponentProps> = ({
  type,
  path,
  overrideProps,
  props,
  update
}) => {
  const { value, label } = props;
  const [val, unit] = stripUnit(value);

  switch (type) {
    case 'colorpicker':
      return (
        <Colorpicker
          label={label}
          value={value}
          onChange={val => update(path, val)}
          {...overrideProps}
        />
      );
    case 'counter':
      return (
        <Counter
          label={label}
          suffix={unit}
          value={parseFloat(val)}
          onChange={(val, suffix) =>
            update(path, suffix ? `${val}${suffix}` : val)
          }
          {...overrideProps}
        />
      );
    case 'range':
      return (
        <Range
          label={props.label}
          suffix={unit}
          value={parseFloat(val)}
          onChange={(val, suffix) =>
            update(path, suffix ? `${val}${suffix}` : val)
          }
          {...overrideProps}
        />
      );
    case 'input':
      return (
        <Input
          label={label}
          value={value}
          onChange={val => update(path, val)}
          {...overrideProps}
        />
      );
    case 'shorthand':
      return (
        <Shorthand
          label={label}
          value={value}
          onChange={val => update(path, val)}
          {...overrideProps}
        />
      );
    case 'switch':
      return (
        <Switch
          label={label}
          value={value}
          onChange={val => update(path, val)}
          {...overrideProps}
        />
      );
    case 'textarea':
      return (
        <Textarea
          label={label}
          value={value}
          onChange={val => update(path, val)}
          {...overrideProps}
        />
      );
    case 'select':
      return (
        <Select
          label={label}
          value={value}
          onChange={val => update(path, val)}
          {...overrideProps}
        />
      );
    case 'radio':
      return (
        <RadioGroup
          label={label}
          value={value}
          name={label}
          onChange={val => update(path, val)}
          {...overrideProps}
        />
      );
    default:
      return null;
  }
};

const areEqual = (prev, next) => {
  const prevOverrideProps = prev.overrideProps ? prev.overrideProps : null;
  const nextOverrideProps = next.overrideProps ? next.overrideProps : null;

  return (
    prev.props.value === next.props.value &&
    JSON.stringify(prevOverrideProps) === JSON.stringify(nextOverrideProps)
  );
};

export const MemoizedSettingsComponent = React.memo(
  SettingsComponent,
  areEqual
);

const SettingsItem = () => {
  const {
    updateTheme,
    overrides,
    config,
    themeComponents,
    activeTheme: { name }
  } = React.useContext(ThemePlaygroundContext);

  const activeComponents = themeComponents[name] || null;

  return (
    <>
      {activeComponents &&
        Object.keys(activeComponents).map(path => {
          const { value, type } = activeComponents[path];
          const label = getLabel(path, config.labelFormat);
          const props = {
            value,
            label
          };

          const componentProps = {
            type,
            path,
            props,
            overrideProps: overrides[path],
            update: updateTheme
          };

          return activeComponents[path] ? (
            <StyledSettingsItem key={path}>
              <MemoizedSettingsComponent {...componentProps} />
            </StyledSettingsItem>
          ) : null;
        })}
    </>
  );
};

export default SettingsItem;
