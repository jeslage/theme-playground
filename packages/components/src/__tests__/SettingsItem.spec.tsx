import 'jest';
import * as React from 'react';
import { shallow } from 'enzyme';

import { StyledSettingsItem } from '../components/SettingsItem/SettingsItem.style';
import { SettingsComponent } from '../components/SettingsItem/SettingsItem';

import { darkTheme } from '../theme';

describe('SettingsItem', () => {
  it('SettingsComponent should render Counter', () => {
    const component = shallow(
      <SettingsComponent
        type="counter"
        path="counter.path"
        overrideProps={{}}
        props={{ label: 'counter.path', value: 20 }}
        update={(path, value) => console.log(path, value)}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('SettingsComponent should render Colorpicker', () => {
    const component = shallow(
      <SettingsComponent
        type="colorpicker"
        path="colorpicker.path"
        overrideProps={{}}
        props={{ label: 'colorpicker.path', value: '#fff' }}
        update={(path, value) => console.log(path, value)}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('SettingsComponent should render Range', () => {
    const component = shallow(
      <SettingsComponent
        type="range"
        path="range.path"
        overrideProps={{}}
        props={{ label: 'range.path', value: 20 }}
        update={(path, value) => console.log(path, value)}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('SettingsComponent should render Input', () => {
    const component = shallow(
      <SettingsComponent
        type="input"
        path="input.path"
        overrideProps={{}}
        props={{ label: 'input.path', value: 'Text' }}
        update={(path, value) => console.log(path, value)}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('SettingsComponent should render Switch', () => {
    const component = shallow(
      <SettingsComponent
        type="switch"
        path="switch.path"
        overrideProps={{}}
        props={{ label: 'switch.path', value: true }}
        update={(path, value) => console.log(path, value)}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('SettingsComponent should render Textarea', () => {
    const component = shallow(
      <SettingsComponent
        type="textarea"
        path="textarea.path"
        overrideProps={{}}
        props={{ label: 'textarea.path', value: 'Textarea value' }}
        update={(path, value) => console.log(path, value)}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('SettingsComponent should render Shorthand', () => {
    const component = shallow(
      <SettingsComponent
        type="shorthand"
        path="shorthand.path"
        overrideProps={{}}
        props={{
          label: 'shorthand.path',
          value: { top: 20, right: 20, left: 20, bottom: 20 }
        }}
        update={(path, value) => console.log(path, value)}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('SettingsComponent should render null', () => {
    const component = shallow(
      <SettingsComponent
        type="unknown"
        path="unknown.path"
        overrideProps={{}}
        props={{
          label: 'unknown.path',
          value: 'not known'
        }}
        update={(path, value) => console.log(path, value)}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('StyledSettingsItem should render correctly', () => {
    const component = shallow(
      <StyledSettingsItem theme={{ colors: darkTheme }} />
    );

    expect(component).toMatchSnapshot();
  });
});
