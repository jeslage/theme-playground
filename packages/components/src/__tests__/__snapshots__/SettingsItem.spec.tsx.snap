// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsItem SettingsComponent should render Colorpicker 1`] = `
<ColorPicker
  label="colorpicker.path"
  onChange={[Function]}
  value="#fff"
/>
`;

exports[`SettingsItem SettingsComponent should render Counter 1`] = `
<Counter
  label="counter.path"
  onChange={[Function]}
  value={20}
/>
`;

exports[`SettingsItem SettingsComponent should render Input 1`] = `
<Input
  label="input.path"
  onChange={[Function]}
  value="Text"
/>
`;

exports[`SettingsItem SettingsComponent should render Range 1`] = `
<Range
  label="range.path"
  onChange={[Function]}
  value={20}
/>
`;

exports[`SettingsItem SettingsComponent should render Shorthand 1`] = `
<Shorthand
  label="shorthand.path"
  onChange={[Function]}
  value={
    Object {
      "bottom": 20,
      "left": 20,
      "right": 20,
      "top": 20,
    }
  }
/>
`;

exports[`SettingsItem SettingsComponent should render Switch 1`] = `
<Switch
  label="switch.path"
  onChange={[Function]}
  value={true}
/>
`;

exports[`SettingsItem SettingsComponent should render Textarea 1`] = `
<Textarea
  label="textarea.path"
  onChange={[Function]}
  value="Textarea value"
/>
`;

exports[`SettingsItem SettingsComponent should render null 1`] = `""`;

exports[`SettingsItem StyledSettingsItem should render correctly 1`] = `
<div
  className="sc-AxjAm bwvmfB"
/>
`;
